python/ray/util/multiprocessing/pool.py

```diff
@@ -384,7 +429,12 @@ class OrderedIMapIterator(IMapIterator):
 
     def next(self, timeout=None):
         if len(self._ready_objects) == 0:
-            if self._next_chunk_index == self._total_chunks:
+            if self._finished_iterating and (
+                self._next_chunk_index == len(self._submitted_chunks)
+            ):
+                # Stop the result_thread and wait for it to complete
+                self._result_thread.stop()
```

Comment:
Does this mean the `_result_thread` will run indefinitely if `next` isn't called at least once after all the elements are consumed?

Reply:
Yes, it does. But within `map()` (the intended application), that's exactly what happens, right? I think one could try to be a bit cleverer, but that would involve putting some logic into `ResultThread` - which cannot know for sure that the `IMapIterator` functions are not still sending new data. As a result I tried to keep it simple: only when all submitted chunks are done *and* we know no more data is coming we can safely shut down the thread.

Reply:
Those are good points for `map()`. I'm also wondering how this affects users of `imap()` and `imap_unordered()`. 

However, when users invoke those with an iterable of unknown length, I think it's a fair assumption that they'll call `next()` until they hit the `StopIteration`. I'm okay with leaving it this way and adding a comment to `next()` saying that it needs to be called one extra time to shut down the thread.

@edoakes what do you think? Should this also be documented in public docstrings somewhere?

Reply:
Sorry, I'm actually thinking of `imap` and `imap_unordered` (which I'm using in my own application - the reason to dive into this). As far as I can tell, these are the only functions instantiating the `IMapIterator` objects and they always exhaust the iterator, so that would not be an issue. 

I can add a comment saying that the `ResultThread` will only be released when the iterators is exhausted (i.e. the `n+1`th invocation of `next`)

Reply:
This issue no longer occurs in the latest version: the `run()` function (and therefore the thread) will exit when the final chunk has been processed. This is possible because the submitting process now sets a flag for the final chunk.